Updating PR 299 (NSURLSession) to work with the latest Foundation

pushkarnk opened the pull request:
Creating this PR for review, on @parkera 's suggestion. The changes that this PR does to pull request 299 include:
1. Rebasing to the latest Foundation
2. Addition of support for download task and upload task (with a completion handler)
3. Enabling building on Ubuntu 14.04 (which has an older libcurl by default)
4. Renaming internal types (classes, struct, enums and protocols) to start with _

Work in progress:
1. Unit tests contributed to TestFoundation as a part of pull request 299 - these utilise types internal to Foundation - don't build on Linux. They need rework.
2. Replacing uses of dispatch_data by NSData/Data backed by dispatch_data

Kindly review and comment. Thanks.

parkera commented on Foundation/NSURLSession/EasyHandle.swift:
```diff
@@ -0,0 +1,632 @@
+         let errorCode: Int?
+     }
+ }
+ */
```
Let's not check in commented out code, since its purpose is unclear.

parkera commented on Foundation/NSURLSession/EasyHandle.swift:
```diff
@@ -0,0 +1,632 @@
+    var connectFailureErrno: CInt {
+        var errno = Int()
+        try! CFURLSession_easy_getinfo_long(rawHandle, CFURLSessionInfoOS_ERRNO, &errno).asError()
+        return CInt(errno)
```
We should strive to make the C interface disappear as soon as possible when importing into Swift - so API like this should probably return `Int` instead of something like `CInt`.

parkera commented on Foundation/NSURLSession/EasyHandle.swift:
```diff
@@ -0,0 +1,632 @@
+        try! CFURLSession_easy_getinfo_charp(rawHandle, CFURLSessionInfoREDIRECT_URL, &p).asError()
+        guard let cstring = p.map({ UnsafePointer<UTF8.CodeUnit>($0) }) else { return nil }
+        guard let s = String.decodeCString(cstring, as: UTF8.self, repairingInvalidCodeUnits: false)?.0 else { return nil }
+        return URL(string: s)
```
String's `init?(bytes:encoding:)` is the right API to use here.

parkera commented on Foundation/NSURLSession/EasyHandle.swift:
```diff
@@ -0,0 +1,632 @@
+        URLSession.printDebug("[EasyHandle] -> write callback \(size * nmemb)")
+        let d: Int = {
+            let buffer = UnsafeBufferPointer(start: data, count: size * nmemb)
+            switch delegate.didReceive(data: buffer) {
```
As discussed earlier, let's try to get these kinds of things into `struct Data` to present a more sane Swift interface after we've received stuff from the C layer.

parkera commented on Foundation/NSURLSession/HTTPBodySource.swift:
```diff
@@ -0,0 +1,261 @@
+internal extension NSData {
+    /// Turn `dispatch_data_t` into `NSData`
+    convenience init(dispatchData: dispatch_data_t) {
+        //TODO: Should do this through an NSData subclass to avoid copying
```
cc @phausler

phausler replied:
Ideally we should make a subclass of NSData that wraps that object. Do we have a swift class in dispatch for OS_dispatch_data?

seabaylea replied:
In the Swift 2.x API for Dispatch we have a `dispatch_data_t` class that wraps the C implementation.

parkera commented on Foundation/NSURLSession/HTTPBodySource.swift:
```diff
@@ -0,0 +1,261 @@
+/// should all happen on the same (serial) queue, and hence this code doesn't
+/// have to be thread safe.
+internal final class _HTTPBodyFileSource {
+    private let fileURL: NSURL
```
This should be struct URL now.

parkera commented on Foundation/NSURLSession/HTTPBodySource.swift:
```diff
@@ -0,0 +1,261 @@
+}
+
+extension _HTTPBodyFileSource : _HTTPBodySource {
+    func getNextChunk(withLength length: Int) -> _HTTPBodySourceDataChunk { //fatalError("unimplemented") }   
```
delete comment

parkera commented on Foundation/NSURLSession/HTTPMessage.swift:
```diff
@@ -0,0 +1,367 @@
+                data[data.endIndex - 1] == LF
+            else { return nil }
+        let lineBuffer = UnsafeBufferPointer(start: UnsafePointer<UInt8>(data.baseAddress), count: data.count - 2)
+        guard let line = String(utf8Buffer: lineBuffer) else { return nil }
```
Should be `init?(bytes:encoding:)` again here (and elsewhere we create strings) from data.

parkera commented on Foundation/NSURLSession/HTTPMessage.swift:
```diff
@@ -0,0 +1,367 @@
+    /// This will parse the header lines.
+    /// - Returns: `nil` if an error occured while parsing the header lines.
+    @warn_unused_result
+    func createHTTPURLResponse(for URL: URL) -> NSHTTPURLResponse? {
```
Hm, did this class not drop its NS?

pushkarnk replied:
Not yet, though it's listed [here](https://github.com/apple/swift-evolution/blob/master/proposals/0086-drop-foundation-ns.md). If there was no reason for not dropping it, I can get it done.

parkera commented on Foundation/NSURLSession/HTTPMessage.swift:
```diff
@@ -0,0 +1,367 @@
+// Characters that we need for HTTP parsing:
+
+/// *Carriage Return* symbol
+private let CR: Int8 = 0x0d
```
Let's at least stick these super generic names inside some kind of scope.

parkera commented on Foundation/NSURLSession/HTTPMessage.swift:
```diff
@@ -0,0 +1,367 @@
+private let HorizontalTab = UnicodeScalar(0x09)
+private let Colon = UnicodeScalar(0x3a)
+/// *Separators* according to RFC 2616
+private let Separators: [UnicodeScalar] = [
```
We should use `CharacterSet` for this kind of thing.

parkera commented on Foundation/NSURLSession/MultiHandle.swift:
```diff
@@ -0,0 +1,429 @@
+    /// easy handle.
+    /// - Note: The error value is set only on failure. You can't use it to
+    ///   determine *if* something failed or not, only *why* it failed.
+    func URLErrorCode(for easyCode: CFURLSessionEasyCode) -> Int? {
```
naming conventions say that this should be `urlErrorCode`.

parkera commented on Foundation/NSURLSession/NSURLSession.swift:
```diff
@@ -0,0 +1,511 @@
+
+/*
+ 
+ NSURLSession is a replacement API for NSURLConnection.  It provides
```
We should update this comment to use the new class names (not a blocker).

pushkarnk commented:
Thanks for the comments. I started with a round of deleting commented code with no message.

pushkarnk commented:
I addressed a few of the comments today:
1. Use of String.init?(bytes:encoding:)  and String.init?(cString:encoding:) to create Strings from char pointers received from the CURL layer
2. Adhering to the new naming conventions
3. Moved some constants in HTTMessage inside a struct
4. Replaced the use of a [UnicodeScalar] by NSCharacterSet

phausler commented on build.py:
```diff
@@ -50,22 +51,25 @@
 		'-L${XCTEST_BUILD_DIR}',
-		'-I${SYSROOT}/usr/include/libxml2'
+		'-I${SYSROOT}/usr/include/libxml2',
+                '-I/usr/include/curl'
```
This should also be `${SYSROOT}` attached just like the libxml case so that way cross compile development can specify a root

phausler commented on Foundation/NSURLSession/libcurlHelpers.swift:
```diff
@@ -0,0 +1,53 @@
+        var bufferIterator = utf8Buffer.makeIterator()
+        var codec = UTF8()
+        var result: String = ""
+        iter: repeat {
```
I am dubious of the actual perf hit of this; (not a blocking issue) I think we can probably do much better than this and funnel it to a common interface in CF

pushkarnk commented:
More changes:
1. Tried to replace CInt with Int but that can't be done everywhere since libcurl accepts and returns Int32s.
2. Replace NSURL with struct URL.

pushkarnk commented:
I have replaced Unsafe pointers with struct Data in all places except for the libcurl callbacks

parkera commented:
Ok - I want to merge this but I have one strict requirement - we need at least the start of a test suite. Is it possible to get some kind of straightforward loop-back test in place so that we know we aren't going to accidentally break this with changes in the future?

pushkarnk commented:
Sure @parkera I will work on getting some of Daniel's tests running.

pushkarnk commented:
@parkera @phausler : Whats the equivalent of Xcode's "Enable Testability" on Linux? All of Daniel's test cases could be added here if we could get the @testable import Foundation to work, since most of his tests use internal classes. I see this error currently: 
```
TestFoundation/TestNSURLSession.swift:17:22: error: module 'Foundation' was not compiled for testing
    @testable import Foundation
```

parkera commented:
Isn't the main purpose of `@testable` to expose some otherwise-private implementation details? Do the tests use those, or do they primarily test the public interface?

pushkarnk commented:
Some tests use the internal classes to test the public API. Other than that, there are tests for the internal classes too. We'd need @testable for both

parkera commented:
IIRC the challenge with `@testable` on Linux was that the build bot would have to build it this way when running its own tests, but we don't want that to be the way it's built for final consumption...

pushkarnk commented:
I see that -enable-testing is the flag.

danieleggert commented:
Just seeing this now. I’ve started rebasing my old PR, too.

pushkarnk commented:
@seabaylea @parkera - As per your suggestions I have done the following:

1. Migrated to the new Dispatch API
2. Added 6 basic tests (4 data task + 2 download task) with external URLs - TestURLSession.swift. I think uploadTask tests need to be done via the loopback way.
3. Given that (2) is a stopgap way of testing URLSession, opened a Jira task requesting for loopback tests contributed from the community - https://bugs.swift.org/browse/SR-2118

I've also rebased to the latest Foundation.

parkera commented:
Awesome. Kicking off a test.

parkera commented:
@swift-ci please test

parkera commented:
Do we have a configuration issue?

```
Foundation/NSURLSession/EasyHandle.swift:21:8: error: no such module 'Dispatch'
import Dispatch
```

seabaylea commented:
Dispatch so far has only been added to the libdispatch CI builds:
https://ci.swift.org/view/All/job/oss-swift-incremental-RA-libdispatch-linux-ubuntu-15_10/
and those are currently failing because the build timeout is too short for the duration of the `check-swift-validation-linux-x86_64` tests.

As well as asking @shahmishal to extend the timeout for the libdispatch builds, we could also ask for libdispatch to be added to the Linux PR builds

pushkarnk commented:
rebased

parkera commented:
Thanks @pushkarnk . I'm still trying to figure out exactly where we are in this.

pushkarnk commented:
rebased to the latest compiler and stdlib

seabaylea commented:
We've just embedded the BlocksRuntime in libdispatch, which should mean that the "Swift Foundation Pull Request Test Linux" CI build is capable is building the NSURLSession work.

Would it be possible for you to ask the CI to kick off a PR test?

pushkarnk commented:
Thanks @seabaylea 

I have rebased again now (included the AnyObject to Any change). Also added implementations for three API methods:
```
    uploadTask(with:fromFile:completionHandler:)
    downloadTask(with:completionHandler:) //with url
    downloadTask(with:completionHandler:) //with request
```

seabaylea commented:
@parkera could you kick off a PR test for this? The "Swift Foundation Pull Request Test Linux" builds should now have the pre-reqs for the build to complete.

pushkarnk commented:
rebased

pushkarnk commented:
Rebased. Also changed some names to match those in the 08-15 toolchain on Darwin.

pushkarnk commented:
Rebased.

pushkarnk commented:
Today, after rebasing I found all changes missing (guess it was due to a commit and a subsequent revert). I wasn't sure about how to recover from that condition. So, I have manually added a new commit with the same changes (and some more related to yesterday's compiler and Foundation changes).

parkera commented:
@swift-ci please test linux

parkera commented:
```
Foundation/NSURLSession/EasyHandle.swift:21:8: error: no such module 'Dispatch'
import Dispatch
```

I saw the checkout in the build script but not the libdispatch build...

seabaylea commented:
@parkera @shahmishal  I've just sent an email to you about this (and some other failures). This is a build configuration issue. The job config needs to be:
```
swift/utils/build-script -R -T --llbuild --swiftpm --xctest --foundation --libdispatch -- --reconfigure
```
ie, `--libdispatch` needs to be added

shahmishal commented:
@swift-ci Please test Linux

parkera commented:
Great. The compiler crashed building swiftpm. =(

https://ci.swift.org/job/swift-corelibs-foundation-PR-Linux/235/consoleFull#979178979ee1a197b-acac-4b17-83cf-a53b95139a76

```
Compile Swift Module 'Basic' (20 sources)
swift: /home/buildnode/disk2/workspace/swift-corelibs-foundation-PR-Linux/swift/include/swift/Serialization/ModuleFile.h:367: Status swift::ModuleFile::error(Status): Assertion `(!FileContext || issue != Status::Malformed) && "error deserializing an individual record"' failed.
```

parkera commented:
On the plus side, URLSession built:

```
[262/314] CompileSwift: Foundation/NSURLResponse.swift
[263/314] CompileSwift: Foundation/NSURLSession/EasyHandle.swift
[264/314] CompileSwift: Foundation/NSURLSession/HTTPBodySource.swift
[265/314] CompileSwift: Foundation/NSURLSession/MultiHandle.swift
[266/314] CompileSwift: Foundation/NSURLSession/HTTPMessage.swift
[267/314] CompileSwift: Foundation/NSURLSession/NSURLSession.swift
[268/314] CompileSwift: Foundation/NSURLSession/NSURLSessionConfiguration.swift
[269/314] CompileSwift: Foundation/NSURLSession/NSURLSessionDelegate.swift
[270/314] CompileSwift: Foundation/NSURLSession/TaskRegistry.swift
[271/314] CompileSwift: Foundation/NSURLSession/NSURLSessionTask.swift
[272/314] CompileSwift: Foundation/NSURLSession/libcurlHelpers.swift
[273/314] CompileSwift: Foundation/NSURLSession/TransferState.swift
```

parkera commented:
I'm just going to kick this off again, just in case.

parkera commented:
@swift-ci please test linux

parkera commented:
Crashed again. It looks like it might be related to dispatch inclusion:

```
1.	While type-checking 'fputs' at /home/buildnode/disk2/workspace/swift-corelibs-foundation-PR-Linux/swiftpm/Sources/Basic/PathShims.swift:274:8
2.	While type-checking expression at [/home/buildnode/disk2/workspace/swift-corelibs-foundation-PR-Linux/swiftpm/Sources/Basic/PathShims.swift:275:5 - line:275:36] RangeText="handle.write(Data(bytes: bytes))"
3.	While loading members for declaration 0x6a06150 at <invalid loc>
4.	While deserializing 'init' (ConstructorDecl #11177) 
5.	While deserializing decl #27528 (PARAM_DECL)
6.	While deserializing decl #34459 (XREF)
7.	Cross-reference to module 'Dispatch'
	... DispatchData
```

parkera commented:
@jrose-apple Jordan, do you have any insight into this assertion?

seabaylea commented:
I can reproduce this locally. The failure only occurs with the NSURLSession code merged, but there's no obvious reason why it would be directly related - there's no direct usage of NSURLSession by the SwiftPM tests (that I'm aware of), and Dispatch is already included (via Foundation) in the builds that I'm doing without the NSURLSession changes.

There's a similar looking bug here:
https://bugs.swift.org/browse/SR-2040
although that's when cross-referencing `Core` rather then `Dispatch`:
```
1.      While reading from /tmp/repro/.build/debug/Crash.build/crash~partial.swiftmodule
2.      While deserializing 'crash' (FuncDecl #20) 
3.      While deserializing decl #53 (XREF)
4.      Cross-reference to module 'Core'
        ... B
        ... generic param #0
```

parkera commented:
cc @slavapestov @rudkx

parkera commented:
Yah it's 100% reproducible for me as well locally, but I can't figure out what the root cause is either. I was thinking that it was related to a typealias of some kind (maybe one for Data?) but I can't find any in swiftpm or XCTest.

rudkx commented:
@parkera Do you know offhand if this will reproduce on macOS if I pull down the changes?

parkera commented:
I don't think so, because I don't believe swift-corelibs-libdispatch actually builds on macOS right now.

seabaylea commented:
The problem looks like it occurs when importing DispatchData in two cases:
1. Where its used as part of an extension to `Data` to provide a new `init()` function that takes a `DispatchData` stuct
2. Where is used as a type to an internal `split` function.

These get imported by the SwiftPM builds as some of the modules use a Data.init() function (not the extension provided by NSURLSession though), and String.split() (not the split defined in NSURLSession though).

Its possible to work around the issues by:
1. Removing the `Data.init()` extension function - it not actually used anywhere
2. Renaming the `split()` function definition and its usage in NSURLSession to something else

This provides a clean build. Additionally I've done some basic tests that use Dispatch and DispatchData directly, so it looks like the problems only there when importing as a cross-reference.

Ideally we should get the compiler failure fixed, but if that's problematic then working around the issues might be an option.

pushkarnk commented:
@seabaylea  : Thank you. I have updated the PR with the work around you suggested. I can see swiftpm build without problems now.

@parkera : can you please kick off a test again? Thanks.

rudkx commented:
@pushkarnk I can start the tests up.

Can you open a bug at bugs.swift.org for the compiler issue so that we can track that down?

rudkx commented:
@swift-ci please test linux

seabaylea commented:
Looks like it passed the compile but failed one of the tests which is set to a 5 second timeout:

```
Test Case 'TestURLSession.test_dataTaskWithURLCompletionHandler' started at 14:10:19.702
TestFoundation/TestNSURLSession.swift:64: error: TestURLSession.test_dataTaskWithURLCompletionHandler : Asynchronous wait failed - Exceeded timeout of 5.0 seconds, with unfulfilled expectations: URL test with completion handler
```
The test makes a connection to `https://restcountries.eu/rest/v1/name/USA?fullText=true` and the test passes for me locally. The previous two tests in the set hit the same URL and passed, but subsequent tests failed - so we may have suffered from an outage at the remote side.

parkera commented:
We're going to need a remote URL that's as close to 100% reliable as possible.

parkera commented:
@swift-ci please test Linux

jrose-apple commented:
This looks like Dispatch and Foundation are out of sync. We need to make sure Dispatch is always built before Foundation is if we're not already doing that.

parkera commented:
@jrose-apple it doesn't seem to be that straightforward; this was certainly reproducing for me while dispatch was built first.

```
libdispatch: using gold linker
Skipping reconfiguration of libdispatch
+ pushd /home/parkera/Development/build/Ninja-RelWithDebInfoAssert/libdispatch-linux-x86_64
~/Development/build/Ninja-RelWithDebInfoAssert/libdispatch-linux-x86_64 ~/Development
+ make
Making all in dispatch
<stuff>
+ popd
~/Development
+ pushd /home/parkera/Development/build/Ninja-RelWithDebInfoAssert/libdispatch-linux-x86_64/tests
~/Development/build/Ninja-RelWithDebInfoAssert/libdispatch-linux-x86_64/tests ~/Development
+ make build-tests
make: Nothing to be done for `build-tests'.
+ popd
~/Development
foundation: using gold linker
+ pushd /home/parkera/Development/swift-corelibs-foundation
~/Development/swift-corelibs-foundation ~/Development
```

seabaylea commented:
@parkera "We're going to need a remote URL that's as close to 100% reliable as possible." - agreed, as well as building in some resiliency to ignore requests that timeout (rather than an actual failure).

parkera commented:
Hm, it failed again

```
Test Case 'TestURLSession.test_dataTaskWithURLRequest' started at 16:00:24.139
TestFoundation/TestNSURLSession.swift:72: error: TestURLSession.test_dataTaskWithURLRequest : Asynchronous wait failed - Exceeded timeout of 5.0 seconds, with unfulfilled expectations: data task
TestFoundation/TestNSURLSession.swift:73: error: TestURLSession.test_dataTaskWithURLRequest : XCTAssertEqual failed: ("unknown") is not equal to ("Lima") - test_dataTaskWithURLRequest returned an unexpected result
```

phausler commented:
I was getting similar failures when I was merging it in previously. IIRC it had to do with the JSON getting emitted differently from the id to Any conversions

jrose-apple commented:
Ah, is the Dispatch overlay also rebuilt before you recompile Foundation? I remember we had odd ordering around that at one point.

parkera commented:
@jrose-apple I'm not sure how to tell. Can you look at the build log and let me know what you think?

seabaylea commented:
@parkera @jrose-apple whilst foundation is in build-presets.ini before libdispatch (I've always done it the other way in local builds), from the CI build log it looks to me like libdispatch, including the overlay, builds first regardless.

seabaylea commented:
If we want to add timeout support, and use that to add either retry logic or ignore the test, it looks like we just need to add the following at line 243 in MultiHandle.swift:
```
        case (CFURLSessionEasyCodeCOULDNT_CONNECT, ETIMEDOUT):
            return NSURLErrorTimedOut
```

jrose-apple commented:
Okay, yeah, it looks like it's building in the right order. Actually, it looks like it's finding the dispatch headers, but *not* the overlay. Maybe that's not in the include paths for swiftpm?

jrose-apple commented:
(The reason the compiler doesn't catch the missing overlay is because I wanted people to feel free to move things between their Swift and C code when it was safe to do so.)

parkera commented:
@pushkarnk Can you update the PR as @seabaylea recommends? Once the test passes reliably we can merge this.

seabaylea commented:
@jrose-apple thanks - that was it. I've raised PRs against SwiftPM and XCTest to add the build-time location of the Swift 3 overlay to their build/test-time include paths:
https://github.com/apple/swift-package-manager/pull/615
https://github.com/apple/swift-corelibs-xctest/pull/169

jrose-apple commented:
How does this work when we build a toolchain package? Is the Dispatch overlay installed next to the standard library as it is on Apple platforms?

seabaylea commented:
For the toolchain we install:
```
./lib/swift/dispatch/dispatch.h
./lib/swift/linux/libdispatch.la
./lib/swift/linux/libdispatch.so
./lib/swift/shims/DispatchShims.h
./lib/swift/linux/x86_64/Dispatch.swiftdoc
./lib/swift/linux/x86_64/Dispatch.swiftmodule
```
which matches the location of the other libraries

pushkarnk commented:
@parkera , @seabaylea : I have added some resiliency in the tests by handling timeouts.  I also had to add a basic implementation for ```URLSessionConfiguration.timeoutIntervalForRequest```

Can we test again please, to see how things go?

parkera commented:
@swift-ci please test linux

parkera commented:
It passed!

pushkarnk commented:
Thanks @parkera 

@phausler  - I'll work on updating the Xcode project next.

pushkarnk commented:
Appears https://github.com/apple/swift-corelibs-foundation/pull/586 does it. Thanks.

modocache commented:
One of the tests added in this pull request, `TestURLSession.test_dataTaskWithURLCompletionHandler`, has begun failing. Apparently when this test was written, `https://restcountries.eu/rest/v1/name/USA?fullText=true` returned "Washington D.C.". It now returns "Washington, D.C.".

/cc @pushkarnk @parkera

parkera commented:
It looks like #426 addresses that.

modocache commented:
Yup, my bad. Sorry for the noise!

parkera commented:
Sorry, it's #590 actually.

tannernelson commented:
Will this PR be included in the `swift-3.0-branch`?

parkera commented:
We're planning on merging master to swift-3.0-branch once things are settled down a bit.
